feat: React Native performance and measurements (#4082)
getsentry/sentry-docs: src/platforms/react-native/performance/instrumentation/automatic-instrumentation.mdx

- **Span(s)**: This feature will add span(s) to any transaction running on the global Sentry scope of your app.

Below are details on the automatic instrumentation features that we provide:

### App Start Instrumentation

_Type: **Measurement, Span**_

The App Start Instrumentation provides insight into how long your application takes to launch. It tracks how long it takes the app from the **earliest native process initialization** until when the **React Native root component mounts**.

<Note>

If you don't [wrap your root component with Sentry](#wrap-your-root-component), the App Start measurement will finish when the JavaScript code is initialized instead of the first component mount.

</Note>

The SDK differentiates between a cold and a warm start, and doesn't track hot starts/resumes.

- **Cold start**: App launched for the first time, after a reboot or update. The app is not in memory and no process exists.
- **Warm start**: App launched at least once, is partially in memory, and no process exists.

### Slow and Frozen Frames

_Type: **Measurement**_

> Unresponsive UI, animation hitches, or just jank annoy users and degrade the user experience. Two measurements to track this unwanted experience are slow frames and frozen frames. A smoothly running app should try to avoid both. That's why the SDK adds these two measurements for the transactions you capture.

Reviewer: ```suggestion
Unresponsive UI and animation hitches annoy users and degrade the user experience. Two measurements to track these types of experiences are slow frames and frozen frames. If you want your app to run smoothly, you should try to avoid both. The SDK adds these two measurements for the transactions you capture.
```

> - **Slow Frames**: Slow frames are captured by the SDK when the app needs more than 16.67 ms for a frame. Typically, a phone or tablet renders with 60 frames per second (fps), though the frame rate can be higher, especially as 120 fps displays become popular. With 60 fps, every frame has 16.67 ms to render; slower than that, and the app has slow frames.

Reviewer: ```suggestion
- **Slow Frames**: Slow frames are captured by the SDK when the app takes more than 16.67 ms to render a frame. Typically, a phone or tablet renders 60 frames per second (fps), though the frame rate can be higher, especially as 120 fps displays become popular. At 60 fps, every frame has 16.67 ms to render; slower than that, and the app has slow frames.
```

> - **Frozen Frames** Frozen frames are UI frames that take longer than 700 ms.

Reviewer: ```suggestion
- **Frozen Frames** Frozen frames are UI frames that take longer than 700 ms to render.
```

### Stall Tracking

_Type: **Measurement**_

> A stall is when the JavaScript event loop takes longer than expected to complete a loop. A stall in your JavaScript code will not just make your UI unresponsive or janky, but also slow down your logic that is contained within JavaScript, essentially slowing everything down altogether and creating a bad experience for your users.

Reviewer: ```suggestion
A stall is when the JavaScript event loop takes longer than expected to complete. A stall in your JavaScript code will not just make your UI unresponsive, but also slow down the logic that is contained within JavaScript. This slows everything down, creating a bad experience for your users.
```

> We track stalls that occur in your React Native app during the time of a transaction and provide you with these values:

Reviewer: ```suggestion
We track stalls that occur in your React Native app during a transaction and provide you with these values:
```

> - **Longest Stall Time**: The time in milliseconds of the longest event loop stall.

Reviewer: ```suggestion
- **Longest Stall Time**: The time, in milliseconds, of the longest event loop stall.
```

> - **Total Stall Time**: The total combined time in milliseconds of all stalls.

Reviewer: ```suggestion
- **Total Stall Time**: The total combined time, in milliseconds, of all stalls.
```

> - **Stall Count**: The total number of stalls that occured during the time of the transaction.

Reviewer: ```suggestion
- **Stall Count**: The total number of stalls that occurred during the transaction.
```

### Fetch/XML Request Instrumentation

_Type: **Span**_

The tracing integration creates a child span for every `XMLHttpRequest` or `fetch` request on the Javascript layer that occurs while those transactions are open. Learn more about [traces, transactions, and spans](/product/sentry-basics/tracing/distributed-tracing/).

## Configuration Options

> To configure the automatic performance instrumentation, you will need to add the `ReactNativeTracing` integration yourself. We provide many reasonable options by default so for the majority of apps you would not need to configure the integration yourself.

Reviewer: ```suggestion
To configure the automatic performance instrumentation, you will need to add the `ReactNativeTracing` integration yourself. We provide many options by default, so for the majority of apps you won't need to configure the integration yourself.
```

```javascript
import * as Sentry from "@sentry/react-native";

Sentry.init({
  dsn: "___PUBLIC_DSN___",

  integrations: [
    new Sentry.ReactNativeTracing({
      tracingOrigins: ["localhost", "my-site-url.com", /^\//],
      // ... other options
    }),
  ],
});
```

For all possible options, see [ReactNativeTracingOptions](https://github.com/getsentry/sentry-react-native/blob/211ec081b6bf8d7d29541afe9d800040f61e3c4e/src/js/tracing/reactnativetracing.ts#L23).

### beforeNavigate

> You can use `beforeNavigate` to modify the transaction from routing instrumentation before the transaction is created. If you prefer not sending the transaction, set `sampled = false`.

Reviewer: ```suggestion
You can use `beforeNavigate` to modify the transaction from routing instrumentation before the transaction is created. If you prefer not to send the transaction, set `sampled = false`.
```

### tracingOrigins

The default value of `tracingOrigins` is `['localhost', /^\//]`. The React Native SDK will attach the `sentry-trace` header to all outgoing XHR/fetch requests whose destination contains a string in the list or matches a regex in the list. If your frontend is making requests to a different domain, you will need to add the domain there to propagate the `sentry-trace` header to the backend services, which is required to link transactions together as part of a single trace. **The `tracingOrigins` option matches against the entire request URL, not just the domain. Using stricter regex to match certain parts of the URL ensures that requests do not unnecessarily have the `sentry-trace` header attached.**

<PlatformContent includePath="performance/tracingOrigins-example" />

You will need to configure your web server CORS to allow the `sentry-trace` header. The configuration might look like `"Access-Control-Allow-Headers: sentry-trace"`, but the configuration depends on your setup. If you do not allow the `sentry-trace` header, the request might be blocked.

### shouldCreateSpanForRequest

> This function can be used to filter out unwanted spans, such as XHR's running health checks or something similar. By default `shouldCreateSpanForRequest` already filters out everything except what was defined in `tracingOrigins`.

Reviewer: ```suggestion
This function can be used to filter out unwanted spans, such as XHR's running health checks or something similar. By default, `shouldCreateSpanForRequest` already filters out everything except what was defined in `tracingOrigins`.
```

  )
}
```

## Opt Out

> If you seek to use tracing without our automatic instrumentation, you can disable it by setting `enableAutoPerformanceTracking` in your Sentry options and remove the `ReactNativeTracing` integration if you added it:

Reviewer: ```suggestion
If you want to use tracing without our automatic instrumentation, you can disable it by setting `enableAutoPerformanceTracking` in your Sentry options and removing the `ReactNativeTracing` integration, if you added it:
```
